{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "40db8652-a889-46b2-94c4-324c56cb67e7",
   "metadata": {},
   "source": [
    "# Annotation of Erdman genome\n",
    "This notebook aims to create a high quality annotation of the Erdman genome. The approach is the following:\n",
    "1. Extract annotation features from the RATT annotation made based on H37Rv.\n",
    "2. Extract annotation features made _de novo_ with Bakta.\n",
    "3. From the Bakta annotations, keep only those non-overlapping with RATT annotation\n",
    "4. Build genbank record\n",
    "5. Manually fix CDS transferred with invalid ORFs, add translations and assign final Erdman locus tag.\n",
    "6. Build feature table\n",
    "7. Run quality control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "id": "a88a6c12-604b-47d2-9fd6-a0b77a8a46ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from io import StringIO\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "import pandas as pd\n",
    "import bioframe as bf\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c1b2d34-9ac0-430c-9934-0ecabfba6f57",
   "metadata": {},
   "source": [
    "## 1. Extract features from RATT annotation\n",
    "Made based on H37Rv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "id": "19769fb8-8579-4a2e-a674-d4d8329802b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fix_embl_id_line(filename):\n",
    "    '''\n",
    "    ratt makes strange ID line in the embl file,\n",
    "    that biopython doesn't like. This is a quick fix to read it.\n",
    "    '''\n",
    "    with open(filename, 'r') as file:\n",
    "        lines = file.readlines()\n",
    "\n",
    "    # new_first_line = 'ID                   Erdmann; ; ; ; ; ; 0 BP.'\n",
    "    new_first_line = 'ID   Erdman; ; circular; genomic DNA; ; ; 0 BP.'\n",
    "    lines[0] = new_first_line + '\\n'\n",
    "    modified_embl = ''.join(lines)\n",
    "    \n",
    "    return StringIO(modified_embl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "d4e9c3f1-c7fd-4662-bc34-b482056a8308",
   "metadata": {},
   "outputs": [],
   "source": [
    "ratt_file = '../../Results/8_RATT_annotation/RATT_AssemblyRepetitive/Erdmann.Assembly.Repetitive.contig_1.final.embl'\n",
    "bakta_file = '../../Results/7_Bakta_annotation/result.gbff'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "7d9f8edd-5a28-4af3-ae18-db8515355d53",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# write ratt embl output to gb\n",
    "ratt_gb = '../../Results/8_RATT_annotation/RATT_AssemblyRepetitive/Erdmann.Assembly.Repetitive.contig_1.final.gb'\n",
    "SeqIO.write(list(SeqIO.parse(fix_embl_id_line(ratt_file), 'embl')), ratt_gb, 'genbank')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "6c13a79b-2001-47cb-b072-496f5ffcb141",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make a dictionary with the genes transferred by ratt\n",
    "# rv_genes_in_erdman[locus_tag] = location\n",
    "\n",
    "rv_genes_in_erdman = {}\n",
    "\n",
    "for record in SeqIO.parse(fix_embl_id_line(ratt_file), 'embl'):\n",
    "    for feature in record.features:\n",
    "        # get only gene features\n",
    "        if feature.type == 'gene':\n",
    "            \n",
    "            # get rv locus tag\n",
    "            rv_locus_tag = feature.qualifiers['locus_tag'][0]\n",
    "\n",
    "            # add location and strand\n",
    "            rv_genes_in_erdman[rv_locus_tag] = {}\n",
    "            rv_genes_in_erdman[rv_locus_tag]['start'] = feature.location.start\n",
    "            rv_genes_in_erdman[rv_locus_tag]['end'] = feature.location.end\n",
    "            rv_genes_in_erdman[rv_locus_tag]['strand'] = feature.location.strand   \n",
    "\n",
    "\n",
    "ratt_features = []\n",
    "\n",
    "for record in SeqIO.parse(fix_embl_id_line(ratt_file), 'embl'):\n",
    "    for feature in record.features:\n",
    "        # ignore these features:\n",
    "        if feature.type not in ['source', 'repeat_region', 'gene', 'mobile_element']:\n",
    "            \n",
    "            # make a new feature with selected information\n",
    "\n",
    "            # get location\n",
    "            new_feature_start = feature.location.start\n",
    "            new_feature_end = feature.location.end\n",
    "            new_feature_strand = feature.location.strand\n",
    "\n",
    "            if new_feature_strand == 1:\n",
    "                new_feature_strand_symbol = '+'\n",
    "            elif new_feature_strand == -1:\n",
    "                new_feature_strand_symbol = '-'\n",
    "\n",
    "            # make dictionary to store qualifiers of feature\n",
    "            new_feature_qualifiers = feature.qualifiers\n",
    "\n",
    "            # add if there's an Rv locus tag\n",
    "            if 'locus_tag' in feature.qualifiers.keys():\n",
    "                rv_id = feature.qualifiers['locus_tag'][0]\n",
    "            else:\n",
    "                rv_id = None\n",
    " \n",
    "            # fix the CDS start\n",
    "            \n",
    "            # if there's a gene annotation with the same rv_id\n",
    "            if rv_id in rv_genes_in_erdman.keys():\n",
    "                # if the strand is +, update the \"start\" with what the rv_genes_in_erdman dictionary says\n",
    "                if new_feature_strand == 1:\n",
    "                    new_feature_start = rv_genes_in_erdman[rv_id]['start']\n",
    "                elif new_feature_strand == -1:\n",
    "                    # if the strand is -, update the \"end\"\n",
    "                    new_feature_end = rv_genes_in_erdman[rv_id]['end']\n",
    "                \n",
    "            # make new feature\n",
    "            new_feature = SeqFeature(FeatureLocation(new_feature_start, \n",
    "                                                     new_feature_end, \n",
    "                                                     strand=new_feature_strand),\n",
    "                                     type=feature.type,\n",
    "                                     qualifiers=new_feature_qualifiers)\n",
    "\n",
    "\n",
    "            feature_line = ['ErdmanSF2024', feature.type, new_feature_start+1, \n",
    "                            new_feature_end, new_feature_strand_symbol, \n",
    "                            rv_id, 'ratt', new_feature]\n",
    "            ratt_features.append(feature_line)\n",
    "\n",
    "df_ratt = pd.DataFrame(ratt_features, columns=['chrom', 'type', 'start', 'end', \n",
    "                                               'strand', 'rv_equivalent', 'source', 'seqrecord'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "id": "26e3487f-7be1-40cb-9956-4c69c703a399",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>rv_equivalent</th>\n",
       "      <th>source</th>\n",
       "      <th>seqrecord</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ErdmanSF2024</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1</td>\n",
       "      <td>1524</td>\n",
       "      <td>+</td>\n",
       "      <td>Rv0001</td>\n",
       "      <td>ratt</td>\n",
       "      <td>(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ErdmanSF2024</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2052</td>\n",
       "      <td>3260</td>\n",
       "      <td>+</td>\n",
       "      <td>Rv0002</td>\n",
       "      <td>ratt</td>\n",
       "      <td>(2051, 2052, 2053, 2054, 2055, 2056, 2057, 205...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ErdmanSF2024</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3280</td>\n",
       "      <td>4437</td>\n",
       "      <td>+</td>\n",
       "      <td>Rv0003</td>\n",
       "      <td>ratt</td>\n",
       "      <td>(3279, 3280, 3281, 3282, 3283, 3284, 3285, 328...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ErdmanSF2024</td>\n",
       "      <td>CDS</td>\n",
       "      <td>4434</td>\n",
       "      <td>4997</td>\n",
       "      <td>+</td>\n",
       "      <td>Rv0004</td>\n",
       "      <td>ratt</td>\n",
       "      <td>(4433, 4434, 4435, 4436, 4437, 4438, 4439, 444...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ErdmanSF2024</td>\n",
       "      <td>CDS</td>\n",
       "      <td>5240</td>\n",
       "      <td>7267</td>\n",
       "      <td>+</td>\n",
       "      <td>Rv0005</td>\n",
       "      <td>ratt</td>\n",
       "      <td>(5239, 5240, 5241, 5242, 5243, 5244, 5245, 524...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          chrom type  start   end strand rv_equivalent source  \\\n",
       "0  ErdmanSF2024  CDS      1  1524      +        Rv0001   ratt   \n",
       "1  ErdmanSF2024  CDS   2052  3260      +        Rv0002   ratt   \n",
       "2  ErdmanSF2024  CDS   3280  4437      +        Rv0003   ratt   \n",
       "3  ErdmanSF2024  CDS   4434  4997      +        Rv0004   ratt   \n",
       "4  ErdmanSF2024  CDS   5240  7267      +        Rv0005   ratt   \n",
       "\n",
       "                                           seqrecord  \n",
       "0  (0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,...  \n",
       "1  (2051, 2052, 2053, 2054, 2055, 2056, 2057, 205...  \n",
       "2  (3279, 3280, 3281, 3282, 3283, 3284, 3285, 328...  \n",
       "3  (4433, 4434, 4435, 4436, 4437, 4438, 4439, 444...  \n",
       "4  (5239, 5240, 5241, 5242, 5243, 5244, 5245, 524...  "
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ratt.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "id": "8be0a721-a0d0-4d6f-bf8b-856b3f65d5ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "seqrecord\n",
       "CDS             3877\n",
       "tRNA              45\n",
       "misc_feature      33\n",
       "ncRNA             20\n",
       "rRNA               3\n",
       "misc_RNA           2\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 185,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ratt['seqrecord'].apply(lambda x: x.type).value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "010eb668-5e6b-4c5a-aa45-30dcce87f9d0",
   "metadata": {},
   "source": [
    "## 2. Extract annotation features made _de novo_ with Bakta."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "id": "838a7cb5-ecec-4fcd-a43a-620a8a8868e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "bakta_features = []\n",
    "\n",
    "for record in SeqIO.parse(bakta_file, 'genbank'):\n",
    "    for feature in record.features:\n",
    "        # ignore these features:\n",
    "        if feature.type not in ['repeat_region', 'gene', 'mobile_element', 'source']:\n",
    "            \n",
    "            # make a new feature with selected information\n",
    "\n",
    "            # get location\n",
    "            new_feature_start = feature.location.start\n",
    "            new_feature_end = feature.location.end\n",
    "            new_feature_strand = feature.location.strand\n",
    "\n",
    "            if new_feature_strand == 1:\n",
    "                new_feature_strand_symbol = '+'\n",
    "            elif new_feature_strand == -1:\n",
    "                new_feature_strand_symbol = '-'\n",
    "\n",
    "            # make dictionary to store qualifiers of feature\n",
    "            new_feature_qualifiers = feature.qualifiers\n",
    "                \n",
    "            # make new feature\n",
    "            new_feature = SeqFeature(FeatureLocation(new_feature_start, \n",
    "                                                     new_feature_end, \n",
    "                                                     strand=new_feature_strand),\n",
    "                                     type=feature.type,\n",
    "                                     qualifiers=new_feature_qualifiers)\n",
    "\n",
    "\n",
    "            feature_line = ['ErdmanSF2024', feature.type, new_feature_start+1, new_feature_end, new_feature_strand_symbol, None, 'bakta', new_feature]\n",
    "            bakta_features.append(feature_line)\n",
    "\n",
    "df_bakta = pd.DataFrame(bakta_features, \n",
    "                       columns=['chrom', 'type', 'start', 'end', 'strand', \n",
    "                                'rv_equivalent', 'source', 'seqrecord'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "id": "e550c79b-da0e-4bff-b22a-30f7850db0d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ls: ../../Results/Bakta_annotation/: No such file or directory\n"
     ]
    }
   ],
   "source": [
    "!ls -1 ../../Results/Bakta_annotation/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "id": "a9dda299-d3ce-4899-872b-4dc3ec2ab92c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>rv_equivalent</th>\n",
       "      <th>source</th>\n",
       "      <th>seqrecord</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ErdmanSF2024</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1</td>\n",
       "      <td>1524</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>bakta</td>\n",
       "      <td>(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ErdmanSF2024</td>\n",
       "      <td>rep_origin</td>\n",
       "      <td>1525</td>\n",
       "      <td>2051</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>bakta</td>\n",
       "      <td>(1524, 1525, 1526, 1527, 1528, 1529, 1530, 153...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ErdmanSF2024</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2052</td>\n",
       "      <td>3260</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>bakta</td>\n",
       "      <td>(2051, 2052, 2053, 2054, 2055, 2056, 2057, 205...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ErdmanSF2024</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3280</td>\n",
       "      <td>4437</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>bakta</td>\n",
       "      <td>(3279, 3280, 3281, 3282, 3283, 3284, 3285, 328...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ErdmanSF2024</td>\n",
       "      <td>CDS</td>\n",
       "      <td>4482</td>\n",
       "      <td>4997</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>bakta</td>\n",
       "      <td>(4481, 4482, 4483, 4484, 4485, 4486, 4487, 448...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          chrom        type  start   end strand rv_equivalent source  \\\n",
       "0  ErdmanSF2024         CDS      1  1524      +          None  bakta   \n",
       "1  ErdmanSF2024  rep_origin   1525  2051      +          None  bakta   \n",
       "2  ErdmanSF2024         CDS   2052  3260      +          None  bakta   \n",
       "3  ErdmanSF2024         CDS   3280  4437      +          None  bakta   \n",
       "4  ErdmanSF2024         CDS   4482  4997      +          None  bakta   \n",
       "\n",
       "                                           seqrecord  \n",
       "0  (0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,...  \n",
       "1  (1524, 1525, 1526, 1527, 1528, 1529, 1530, 153...  \n",
       "2  (2051, 2052, 2053, 2054, 2055, 2056, 2057, 205...  \n",
       "3  (3279, 3280, 3281, 3282, 3283, 3284, 3285, 328...  \n",
       "4  (4481, 4482, 4483, 4484, 4485, 4486, 4487, 448...  "
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_bakta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "id": "776491bb-8146-4d28-a047-3aba2ef5e10c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "seqrecord\n",
       "CDS           4070\n",
       "tRNA            46\n",
       "ncRNA           23\n",
       "regulatory      13\n",
       "rRNA             3\n",
       "rep_origin       1\n",
       "tmRNA            1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 189,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_bakta['seqrecord'].apply(lambda x: x.type).value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1466a420-727b-4aa7-8131-a36ba7dd4c59",
   "metadata": {},
   "source": [
    "## 3. From the Bakta annotations, keep only those non-overlapping with RATT annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "id": "51fe43ef-f4f0-417e-9ff3-a942e1b6df3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "overlaps = bf.count_overlaps(df_bakta, df_ratt)\n",
    "to_transfer_from_bakta = overlaps[overlaps['count'] == 0]\n",
    "\n",
    "df_all_features = pd.concat([df_ratt, to_transfer_from_bakta]).sort_values('start')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "id": "4e7bbc04-254c-44ad-8c72-dff46575229a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>rv_equivalent</th>\n",
       "      <th>source</th>\n",
       "      <th>seqrecord</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ErdmanSF2024</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1</td>\n",
       "      <td>1524</td>\n",
       "      <td>+</td>\n",
       "      <td>Rv0001</td>\n",
       "      <td>ratt</td>\n",
       "      <td>(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ErdmanSF2024</td>\n",
       "      <td>rep_origin</td>\n",
       "      <td>1525</td>\n",
       "      <td>2051</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>bakta</td>\n",
       "      <td>(1524, 1525, 1526, 1527, 1528, 1529, 1530, 153...</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ErdmanSF2024</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2052</td>\n",
       "      <td>3260</td>\n",
       "      <td>+</td>\n",
       "      <td>Rv0002</td>\n",
       "      <td>ratt</td>\n",
       "      <td>(2051, 2052, 2053, 2054, 2055, 2056, 2057, 205...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ErdmanSF2024</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3280</td>\n",
       "      <td>4437</td>\n",
       "      <td>+</td>\n",
       "      <td>Rv0003</td>\n",
       "      <td>ratt</td>\n",
       "      <td>(3279, 3280, 3281, 3282, 3283, 3284, 3285, 328...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ErdmanSF2024</td>\n",
       "      <td>CDS</td>\n",
       "      <td>4434</td>\n",
       "      <td>4997</td>\n",
       "      <td>+</td>\n",
       "      <td>Rv0004</td>\n",
       "      <td>ratt</td>\n",
       "      <td>(4433, 4434, 4435, 4436, 4437, 4438, 4439, 444...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          chrom        type  start   end strand rv_equivalent source  \\\n",
       "0  ErdmanSF2024         CDS      1  1524      +        Rv0001   ratt   \n",
       "1  ErdmanSF2024  rep_origin   1525  2051      +          None  bakta   \n",
       "1  ErdmanSF2024         CDS   2052  3260      +        Rv0002   ratt   \n",
       "2  ErdmanSF2024         CDS   3280  4437      +        Rv0003   ratt   \n",
       "3  ErdmanSF2024         CDS   4434  4997      +        Rv0004   ratt   \n",
       "\n",
       "                                           seqrecord  count  \n",
       "0  (0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,...    NaN  \n",
       "1  (1524, 1525, 1526, 1527, 1528, 1529, 1530, 153...    0.0  \n",
       "1  (2051, 2052, 2053, 2054, 2055, 2056, 2057, 205...    NaN  \n",
       "2  (3279, 3280, 3281, 3282, 3283, 3284, 3285, 328...    NaN  \n",
       "3  (4433, 4434, 4435, 4436, 4437, 4438, 4439, 444...    NaN  "
      ]
     },
     "execution_count": 191,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all_features.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "id": "dcadf3ef-eb20-4cbf-97cf-c91aeed7744c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "type\n",
       "CDS             4025\n",
       "tRNA              46\n",
       "misc_feature      33\n",
       "ncRNA             22\n",
       "regulatory         6\n",
       "rRNA               3\n",
       "misc_RNA           2\n",
       "rep_origin         1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 192,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all_features['type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6d766f5-ac48-450c-b97d-83617b8b4967",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "4bb0f5df-2320-4c9b-97dd-f064d0aca47e",
   "metadata": {},
   "source": [
    "## 4. Build genbank record\n",
    "This describes the heuristics of what to transfer from the ratt and Bakta annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "id": "b017b2d0-d30c-407b-aa6c-565ec935a7e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 1\n",
    "new_features = []\n",
    "gene_column = []\n",
    "erd_locus_tag_column = []\n",
    "\n",
    "for column, row in df_all_features.iterrows():\n",
    "    seqrecord = row['seqrecord']\n",
    "    qualifiers = {}\n",
    "    \n",
    "    if row['type']  == 'CDS':\n",
    "        # locus_tag\n",
    "        erdman_locus_tag = f'tmp_Erdman_{str(i).zfill(4)}'\n",
    "        qualifiers['locus_tag'] = erdman_locus_tag\n",
    "        i += 1\n",
    "\n",
    "        # gene\n",
    "        if 'gene' in seqrecord.qualifiers:\n",
    "            qualifiers['gene'] = seqrecord.qualifiers['gene']\n",
    "        elif row['source'] == 'ratt':\n",
    "            if 'locus_tag' in seqrecord.qualifiers:\n",
    "                rv_locus_tag = seqrecord.qualifiers['locus_tag'][0]\n",
    "                qualifiers['gene'] = [f'E{rv_locus_tag}']\n",
    "        else:\n",
    "            qualifiers['gene'] = [erdman_locus_tag]\n",
    "\n",
    "        # product\n",
    "        if 'product' in seqrecord.qualifiers:\n",
    "            qualifiers['product'] = seqrecord.qualifiers['product']\n",
    "\n",
    "        # codon_start\n",
    "        if 'codon_start' in seqrecord.qualifiers:\n",
    "            qualifiers['codon_start'] = seqrecord.qualifiers['codon_start']\n",
    "            \n",
    "        # transl_table\n",
    "        if 'transl_table' in seqrecord.qualifiers:\n",
    "            qualifiers['transl_table'] = seqrecord.qualifiers['transl_table']\n",
    "            \n",
    "        # db_xref\n",
    "        # if row['source'] == 'bakta':\n",
    "        #     if 'db_xref' in seqrecord.qualifiers:\n",
    "        #         qualifiers['db_xref'] = seqrecord.qualifiers['db_xref']\n",
    "\n",
    "        # inference\n",
    "        if row['source'] == 'ratt':\n",
    "            qualifiers['inference'] = 'alignment:ratt:1.0.3:RefSeq:NC_000962.3'\n",
    "        elif row['source'] == 'bakta':\n",
    "            if 'inference' in seqrecord.qualifiers:\n",
    "                qualifiers['inference'] = seqrecord.qualifiers['inference']\n",
    "\n",
    "        # note\n",
    "        if row['source'] == 'ratt':\n",
    "            if 'locus_tag' in seqrecord.qualifiers:\n",
    "                rv_locus_tag = seqrecord.qualifiers['locus_tag'][0]\n",
    "                qualifiers['note'] = f'Simliar to {rv_locus_tag} of M. tuberculosis H37Rv'\n",
    "        elif row['source'] == 'bakta':\n",
    "            qualifiers['note'] = 'De novo annotation with bakta'\n",
    "\n",
    "    elif row['type'] in ['tRNA', 'ncRNA', 'rRNA', 'misc_RNA']:\n",
    "        # locus_tag\n",
    "        erdman_locus_tag = f'tmp_Erdman_{str(i).zfill(4)}'\n",
    "        qualifiers['locus_tag'] = erdman_locus_tag\n",
    "        i += 1\n",
    "        \n",
    "        # gene\n",
    "        if 'gene' in seqrecord.qualifiers:\n",
    "            qualifiers['gene'] = seqrecord.qualifiers['gene']\n",
    "        elif row['source'] == 'ratt':\n",
    "            #print(\"\")\n",
    "            if 'locus_tag' in seqrecord.qualifiers:\n",
    "                qualifiers['gene'] = None\n",
    "            #if 'locus_tag' in seqrecord.qualifiers:\n",
    "            #    rv_locus_tag = seqrecord.qualifiers['locus_tag'][0]\n",
    "            #    qualifiers['gene'] = [f'E{rv_locus_tag}']\n",
    "        else:\n",
    "            print(\"\")\n",
    "            qualifiers['gene'] = [erdman_locus_tag]\n",
    "\n",
    "        # product\n",
    "        if 'product' in seqrecord.qualifiers:\n",
    "            qualifiers['product'] = seqrecord.qualifiers['product']\n",
    "\n",
    "        # db_xref\n",
    "        # if row['source'] == 'bakta':\n",
    "        #     if 'db_xref' in seqrecord.qualifiers:\n",
    "        #         qualifiers['db_xref'] = seqrecord.qualifiers['db_xref']\n",
    "\n",
    "        # inference\n",
    "        if row['source'] == 'ratt':\n",
    "            qualifiers['inference'] = 'alignment:ratt:1.0.3:RefSeq:NC_000962.3'\n",
    "        elif row['source'] == 'bakta':\n",
    "            if 'inference' in seqrecord.qualifiers:\n",
    "                qualifiers['inference'] = seqrecord.qualifiers['inference']\n",
    "\n",
    "        # note\n",
    "        if row['source'] == 'ratt':\n",
    "            if 'locus_tag' in seqrecord.qualifiers:\n",
    "                rv_locus_tag = seqrecord.qualifiers['locus_tag'][0]\n",
    "                qualifiers['note'] = f'Simliar to {rv_locus_tag} of M. tuberculosis H37Rv'\n",
    "        elif row['source'] == 'bakta':\n",
    "            qualifiers['note'] = 'De novo annotation with bakta'\n",
    "\n",
    "        # special transfer only for ncRNA\n",
    "        if row['type'] == 'ncRNA':\n",
    "            qualifiers['ncRNA_class'] = seqrecord.qualifiers['ncRNA_class']\n",
    "\n",
    "    elif row['type'] == 'misc_feature':\n",
    "        # some are in reality pseudogenes\n",
    "        if 'pseudogene' in seqrecord.qualifiers:\n",
    "            seqrecord.type = 'CDS'\n",
    "            \n",
    "            # locus_tag\n",
    "            erdman_locus_tag = f'tmp_Erdman_{str(i).zfill(4)}'\n",
    "            qualifiers['locus_tag'] = erdman_locus_tag\n",
    "            i += 1\n",
    "            \n",
    "            # gene\n",
    "            if 'gene' in seqrecord.qualifiers:\n",
    "                qualifiers['gene'] = seqrecord.qualifiers['gene']\n",
    "            elif row['source'] == 'ratt':\n",
    "                if 'locus_tag' in seqrecord.qualifiers:\n",
    "                    rv_locus_tag = seqrecord.qualifiers['locus_tag'][0]\n",
    "                    qualifiers['gene'] = [f'E{rv_locus_tag}']\n",
    "            else:\n",
    "                qualifiers['gene'] = [erdman_locus_tag]\n",
    "\n",
    "            # product\n",
    "            if 'note' in seqrecord.qualifiers:\n",
    "                qualifiers['product'] =  seqrecord.qualifiers['note'][0]\n",
    "\n",
    "            # note\n",
    "            if row['source'] == 'ratt':\n",
    "                if 'locus_tag' in seqrecord.qualifiers:\n",
    "                    rv_locus_tag = seqrecord.qualifiers['locus_tag'][0]\n",
    "                    qualifiers['note'] = f'Simliar to {rv_locus_tag} of M. tuberculosis H37Rv'\n",
    "            elif row['source'] == 'bakta':\n",
    "                qualifiers['note'] = 'De novo annotation with bakta'\n",
    "\n",
    "            qualifiers['pseudo'] = None\n",
    "            # qualifiers['pseudogene'] = 'unknown'\n",
    "        else:\n",
    "            seqrecord = None\n",
    "        \n",
    "    elif row['type'] == 'regulatory':\n",
    "        # regulatory_class\n",
    "        if 'regulatory_class' in seqrecord.qualifiers:\n",
    "            if row['source'] == 'bakta':\n",
    "                qualifiers['regulatory_class'] = seqrecord.qualifiers['regulatory_class']\n",
    "\n",
    "        # db_xref\n",
    "        # if row['source'] == 'bakta':\n",
    "        #     if 'db_xref' in seqrecord.qualifiers:\n",
    "        #         qualifiers['db_xref'] = seqrecord.qualifiers['db_xref']\n",
    "\n",
    "        # inference\n",
    "        if row['source'] == 'ratt':\n",
    "            qualifiers['inference'] = 'alignment:ratt:1.0.3:RefSeq:NC_000962.3'\n",
    "        elif row['source'] == 'bakta':\n",
    "            if 'inference' in seqrecord.qualifiers:\n",
    "                qualifiers['inference'] = seqrecord.qualifiers['inference']\n",
    "\n",
    "        # function\n",
    "        if row['source'] == 'bakta':\n",
    "            if 'note' in seqrecord.qualifiers:\n",
    "                qualifiers['function'] = seqrecord.qualifiers['note']\n",
    "\n",
    "        # note\n",
    "        if row['source'] == 'ratt':\n",
    "            if 'locus_tag' in seqrecord.qualifiers:\n",
    "                rv_locus_tag = seqrecord.qualifiers['locus_tag'][0]\n",
    "                qualifiers['note'] = f'Simliar to {rv_locus_tag} of M. tuberculosis H37Rv'\n",
    "        elif row['source'] == 'bakta':\n",
    "            qualifiers['note'] = 'De novo annotation with bakta'\n",
    "\n",
    "    elif row['type'] == 'rep_origin':\n",
    "        # function\n",
    "        if row['source'] == 'bakta':\n",
    "            if 'note' in seqrecord.qualifiers:\n",
    "                qualifiers['function'] = seqrecord.qualifiers['note']\n",
    "\n",
    "        # note\n",
    "        if row['source'] == 'bakta':\n",
    "            qualifiers['note'] = 'De novo annotation with bakta'\n",
    "\n",
    "    # make new feature\n",
    "    if seqrecord:\n",
    "        new_feature = SeqFeature(seqrecord.location,\n",
    "                                 type=seqrecord.type,\n",
    "                                 qualifiers=qualifiers)\n",
    "    \n",
    "        new_features.append(new_feature)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "##### Manually adding barcodng plasmid features #####\n",
    "\n",
    "# Here I will manually add the barcoding plasmid features which range from 2764911 to 2770126\n",
    "\n",
    "# 1) general annotation of barcoding plasmid region\n",
    "pjeb_barcoding_vector_Feat = SeqFeature(\n",
    "                        location=FeatureLocation(2764911, 2770126, strand=+1),  # Specify the start and end positions\n",
    "                        type=\"misc_feature\",  # Specify the feature type\n",
    "                        qualifiers={\"label\" : \"pjeb_barcoding_vector\",\n",
    "                                    \"note\": \"Inserted barcoding plasmid region\"}  # Name the feature\n",
    ")\n",
    "\n",
    "new_features.append(pjeb_barcoding_vector_Feat)\n",
    "\n",
    "# 2) attP_core_sequence\n",
    "\n",
    "attP_core_sequence_Feat = SeqFeature(location=FeatureLocation(2764912, 2764955, strand=+1),  # Specify the start and end positions\n",
    "                                type=\"misc_feature\",  # Specify the feature type\n",
    "                                qualifiers={\"label\" : \"attP_core_sequence\",\n",
    "                                            \"note\": \"barcoding plasmid feature - attP_core_sequence\"}  # Name the feature\n",
    "                            )\n",
    "\n",
    "new_features.append(attP_core_sequence_Feat)\n",
    "\n",
    "\n",
    "# 3) qTag_v27\n",
    "qTag_v27_Feat = SeqFeature(location=FeatureLocation(2765549, 2765720, strand=+1),  # Specify the start and end positions\n",
    "                                type=\"misc_feature\",  # Specify the feature type\n",
    "                                qualifiers={\"label\" : \"qTag_v27\",\n",
    "                                            \"note\": \"barcoding plasmid feature - qTag_v27\"}  # Name the feature\n",
    "                            )\n",
    "\n",
    "new_features.append(qTag_v27_Feat)\n",
    "\n",
    "# 4) OriE\n",
    "OriE_Feat = SeqFeature(location=FeatureLocation(2765750, 2766543, strand=+1),  # Specify the start and end positions\n",
    "                                type=\"misc_feature\",  # Specify the feature type\n",
    "                                qualifiers={\"label\" : \"OriE\",\n",
    "                                            \"note\": \"barcoding plasmid feature - OriE\"}  # Name the feature\n",
    "                            )\n",
    "\n",
    "new_features.append(OriE_Feat)\n",
    "\n",
    "\n",
    "# 5) barcode_random_7mer\n",
    "\n",
    "barcode_random_7mer_Feat = SeqFeature(location=FeatureLocation(2765720, 2765727, strand=+1),  # Specify the start and end positions\n",
    "                                type=\"misc_feature\",  # Specify the feature type\n",
    "                                qualifiers={\"label\" : \"barcode_random_7mer\",\n",
    "                                            \"note\": \"barcoding plasmid feature - barcode_random_7mer\"}  # Name the feature\n",
    "                            )\n",
    "\n",
    "new_features.append(barcode_random_7mer_Feat)\n",
    "\n",
    "#################################\n",
    "\n",
    "\n",
    "\n",
    "new_record = list(SeqIO.parse(fix_embl_id_line(ratt_file), 'embl'))[0]\n",
    "new_record.id = 'ErdmanSF2024'\n",
    "new_record.name = 'ErdmanSF2024'\n",
    "\n",
    "source_feature = new_record.features[0]\n",
    "\n",
    "source_feature.qualifiers = {'strain': ['Erdman'],\n",
    "                             'mol_type': ['genomic DNA'],\n",
    "                             'db_xref': ['taxon:652616'],\n",
    "                             'organism': ['Mycobacterium tuberculosis strain Erdman']}\n",
    "new_features.insert(0, source_feature)\n",
    "\n",
    "new_record.features = new_features"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b72b1c6-1a34-4d6e-8d49-ccd50a399a09",
   "metadata": {},
   "source": [
    "## 5. Manually fix CDS transferred with invalid ORF.\n",
    "The file 'manual_fix.tsv' was made by manually inspecting cases of invalid orfs and by fetching the right coordinates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "id": "2304aa4a-c87f-4ac1-a95b-722ddca511da",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_fix = pd.read_csv('manual_fix.tsv',\n",
    "                     sep='\\t')\n",
    "\n",
    "\n",
    "df_fix[['start', 'end']] = df_fix[['start', 'end']].fillna(0).astype(int)\n",
    "\n",
    "fix_dict = {}\n",
    "for column, row in df_fix.iterrows():\n",
    "    gene = row['gene']\n",
    "    fix_dict[gene] = {\n",
    "        'position': (row['start'], row['end']),\n",
    "        'action': row['action'].split(';'),\n",
    "        'note': row['note'] if row['note'] != 'NaN' else None\n",
    "    }\n",
    "\n",
    "\n",
    "clean_features = []\n",
    "\n",
    "# clean features according to what's specified in 'action'\n",
    "for feature in new_record.features:\n",
    "    if 'gene' in feature.qualifiers:\n",
    "        gene = feature.qualifiers['gene'][0]\n",
    "        if gene in fix_dict.keys():\n",
    "            for task in fix_dict[gene]['action']:\n",
    "                if task == 'update_position':\n",
    "                    feature.location = FeatureLocation(fix_dict[gene]['position'][0]-1,\n",
    "                                                       fix_dict[gene]['position'][1],\n",
    "                                                       strand=feature.location.strand)\n",
    "                if task == 'change_strand':\n",
    "                    if feature.location.strand:\n",
    "                        feature.location.strand = -1\n",
    "                    else:\n",
    "                        feature.location.strand = 1\n",
    "                elif task == 'add_note':\n",
    "                    feature.qualifiers['note'] += fix_dict[gene]['note']\n",
    "                elif task == 'pseudo':\n",
    "                    feature.qualifiers['pseudo'] = ''\n",
    "                elif task == 'delete':\n",
    "                    feature = None\n",
    "\n",
    "    # Rerun all DELETE and position-update tasks using the locus tag as well\n",
    "    if feature:\n",
    "        if 'locus_tag' in feature.qualifiers:\n",
    "            i_locus_tag = feature.qualifiers['locus_tag']\n",
    "            if i_locus_tag in fix_dict.keys() and i_locus_tag != gene:\n",
    "                for task in fix_dict[i_locus_tag]['action']:\n",
    "                    if task == 'delete':\n",
    "                        feature = None\n",
    "                    if task == 'update_position':\n",
    "                        feature.location = FeatureLocation(fix_dict[i_locus_tag]['position'][0]-1,\n",
    "                                                           fix_dict[i_locus_tag]['position'][1],\n",
    "                                                           strand=feature.location.strand)\n",
    "\n",
    "    # get translation   \n",
    "    if feature:\n",
    "        if feature.type == 'CDS':\n",
    "            if 'pseudo' not in feature.qualifiers:\n",
    "                if feature.location.strand > 0:\n",
    "                    subseq = new_record.seq[feature.location.start:feature.location.end]\n",
    "                    aa = subseq.translate()\n",
    "                else:\n",
    "                    subseq = new_record.seq[feature.location.start:feature.location.end]\n",
    "                    aa = subseq.reverse_complement().translate()\n",
    "                feature.qualifiers['translation'] = Seq(aa)\n",
    "        \n",
    "        # clean_features.append(feature)\n",
    "\n",
    "\n",
    "\n",
    "    #### Add a label for KanR gene and L5 Integrase\n",
    "\n",
    "        if 'gene' in feature.qualifiers:\n",
    "            gene_name = feature.qualifiers[\"gene\"][0]\n",
    "            if gene_name == \"aph(3')-Ia\":\n",
    "                feature.qualifiers[\"label\"] = [\"aph_KanR\"]\n",
    "                feature.qualifiers[\"note\"] = ['Barcoding plasmid: aph_KanR, kanamycin resistance gene']\n",
    "\n",
    "            if gene_name == \"tmp_Erdman_2579\":\n",
    "                feature.qualifiers[\"label\"] = [\"L5 integrase\"]\n",
    "                feature.qualifiers['gene'][0] = 'L5-gp33'\n",
    "                feature.qualifiers[\"product\"][0] = \"site specific integrase\"\n",
    "                feature.qualifiers[\"note\"] = [\"Barcoding plasmid: site specific integrase (L5 gp33) (UniProt: P22884)\"]\n",
    "    \n",
    "            if gene_name == \"tmp_Erdman_2578\":\n",
    "                feature.qualifiers[\"label\"] = [\"L5 transcriptional repressor\"]\n",
    "                feature.qualifiers['gene'][0] = 'L5-gp71'\n",
    "                feature.qualifiers[\"product\"][0] = \"L5 transcriptional repressor\"\n",
    "                feature.qualifiers[\"note\"] = [\"Barcoding plasmid: L5 transcriptional repressor (L5 gp71) (UniProt: Q05286)\"]\n",
    "\n",
    "        clean_features.append(feature)\n",
    "\n",
    "\n",
    "# RESORT THE clean_features LIST\n",
    "\n",
    "clean_features = sorted(clean_features, key=lambda feature: feature.location.start)\n",
    "\n",
    "# add updated locus_tag\n",
    "i = 1\n",
    "\n",
    "for feature in clean_features:\n",
    "    if 'locus_tag' in feature.qualifiers:\n",
    "        erdman_locus_tag = f'ErdFL_{str(i).zfill(4)}'\n",
    "\n",
    "        feature.qualifiers['locus_tag'] = erdman_locus_tag\n",
    "        i += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "id": "f3e90882-e1d2-4054-ba58-73386dffa002",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4099"
      ]
     },
     "execution_count": 195,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(clean_features)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4cdd32f-1337-43e7-830d-556ba3e3e4f7",
   "metadata": {},
   "source": [
    "## 5.2 Add matching gene features for CDS and tRNA types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "id": "940d3776-98a1-4b3c-82f5-fc6b74f10ba6",
   "metadata": {},
   "outputs": [],
   "source": [
    "clean_features_WiGeneFeats = []\n",
    "\n",
    "for feature in clean_features:\n",
    "\n",
    "    # Retrieve qualifiers (if they exist)\n",
    "    if feature.type in [\"CDS\", 'tRNA', 'ncRNA', 'rRNA', 'misc_RNA']:\n",
    "        locus_tag = feature.qualifiers['locus_tag']\n",
    "        gene = feature.qualifiers.get('gene', [''])[0]\n",
    "        note = feature.qualifiers.get('gene', [''])[0]\n",
    "        if gene == ['']: gene = ''\n",
    "            \n",
    "        # Create a new SeqFeature of type \"gene\"\n",
    "        new_gene_feature = SeqFeature(\n",
    "                location=feature.location,  # Reuse the same location\n",
    "                type=\"gene\",\n",
    "                qualifiers={\n",
    "                    'locus_tag': locus_tag,\n",
    "                    'gene': [gene],\n",
    "                }\n",
    "            )\n",
    "\n",
    "        if 'note' in feature.qualifiers:\n",
    "            new_gene_feature.qualifiers[\"note\"] = feature.qualifiers[\"note\"]\n",
    "\n",
    "        \n",
    "        clean_features_WiGeneFeats.append(new_gene_feature)\n",
    "\n",
    "    if 'gene' in feature.qualifiers:\n",
    "        gene = feature.qualifiers['gene'][0]\n",
    "        \n",
    "        if 'tmp_Erdman' in gene:\n",
    "            del feature.qualifiers['gene']\n",
    "            del new_gene_feature.qualifiers['gene']\n",
    "            gene = locus_tag # None\n",
    "\n",
    "        if 'ERv' in gene:\n",
    "            del feature.qualifiers['gene']\n",
    "            del new_gene_feature.qualifiers['gene']\n",
    "\n",
    "        if gene == \"tnp\":\n",
    "            del feature.qualifiers['gene']\n",
    "            del new_gene_feature.qualifiers['gene']\n",
    "\n",
    "        if gene == \"hypB\":\n",
    "            del feature.qualifiers['gene']\n",
    "            del new_gene_feature.qualifiers['gene']\n",
    "\n",
    "\n",
    "    \n",
    "    clean_features_WiGeneFeats.append(feature)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f68fd4ba-d826-4025-ba17-2962039796ed",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5e47b70e-c4ac-4d53-8662-750f0e660679",
   "metadata": {},
   "source": [
    "## 6. Build feature table and rename `gene` to `locus_tag`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "id": "e0f62357-eb23-42ba-b00e-97aaf0b5918c",
   "metadata": {},
   "outputs": [],
   "source": [
    "annotations = []\n",
    "\n",
    "for feature in clean_features_WiGeneFeats:\n",
    "    if 'locus_tag' in feature.qualifiers:\n",
    "        locus_tag = feature.qualifiers['locus_tag']\n",
    "    else:\n",
    "        locus_tag = None\n",
    "    if 'gene' in feature.qualifiers:\n",
    "        gene = feature.qualifiers['gene'][0]\n",
    "        \n",
    "        # if 'tmp_Erdman' in gene:\n",
    "        #     #gene = locus_tag\n",
    "        #     #feature.qualifiers['gene'][0] = locus_tag\n",
    "        #     del feature.qualifiers['gene']\n",
    "        #     gene = locus_tag # None\n",
    "\n",
    "        # if 'ERv' in gene:\n",
    "        #     del feature.qualifiers['gene']\n",
    "        # if gene == \"tnp\":\n",
    "        #     del feature.qualifiers['gene']\n",
    "\n",
    "    \n",
    "    else:\n",
    "        gene = locus_tag\n",
    "        #gene = None\n",
    "    \n",
    "    record_type = feature.type\n",
    "    start = feature.location.start\n",
    "    end = feature.location.end\n",
    "    if feature.location.strand > 0:\n",
    "        strand = '+'\n",
    "    else:\n",
    "        strand = '-'\n",
    "    \n",
    "    if 'note' in feature.qualifiers:\n",
    "        note = feature.qualifiers['note']\n",
    "        if 'bakta' in note:\n",
    "            source = 'bakta'\n",
    "            rv_equivalent = None\n",
    "            \n",
    "        elif 'H37Rv' in note:\n",
    "            source = 'ratt'\n",
    "            rv_equivalent = note.split('Simliar to ')[1].split(' of M. tuberculosis H37Rv')[0]\n",
    "            \n",
    "    else:\n",
    "        rv_equivalent = None\n",
    "        source = None\n",
    "        note = \"\"\n",
    "    line = [locus_tag, gene, record_type, rv_equivalent, start, end, strand, source, note]\n",
    "    annotations.append(line)\n",
    "\n",
    "df_annotations = pd.DataFrame(annotations, columns = ['locus_tag', 'gene', 'type', 'rv_equivalent', 'start', 'end', 'strand', 'source', 'note'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58910d56-daa6-4953-bf43-9a8f027285f0",
   "metadata": {},
   "source": [
    "### Output curated features TSV and Genbank files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "id": "b0f5cb38-298d-4860-bbf2-9727ff2b13b3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: ./CuratedAnnotationV2: File exists\n"
     ]
    }
   ],
   "source": [
    "!mkdir ./CuratedAnnotationV2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "id": "5b0b0cd3-d75f-4c8f-a2d9-d4e2db66e55e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_annotations.to_csv('./CuratedAnnotationV2/Erdman.curated.V2.tsv', sep='\\t', index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "id": "c513f14d-ba14-49cc-8d4c-945b48fa4f90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# save Genbank\n",
    "new_record.features = clean_features_WiGeneFeats\n",
    "SeqIO.write(new_record, './CuratedAnnotationV2/Erdman.curated.V2.gb', 'genbank')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "808788c4-b7fb-4066-a5d7-a7b17e8c95c3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "id": "95c0bd77-bc47-4772-90dc-0fe5e8e25c04",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "type\n",
       "gene            4086\n",
       "CDS             4014\n",
       "tRNA              46\n",
       "ncRNA             21\n",
       "regulatory         6\n",
       "misc_feature       5\n",
       "rRNA               3\n",
       "misc_RNA           2\n",
       "source             1\n",
       "rep_origin         1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 201,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_annotations[\"type\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23a73718-fe52-40ab-b946-a38650062930",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "61049422-d293-4c07-a2ac-f69fc5cef274",
   "metadata": {},
   "source": [
    "# 5. Quality control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "id": "887234bd-ef42-4a28-a488-7770d730fc79",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start;\n",
      "type: CDS\n",
      "location: [550979:551153](-)\n",
      "qualifiers:\n",
      "    Key: codon_start, Value: ['1']\n",
      "    Key: gene, Value: ['mazE1']\n",
      "    Key: inference, Value: alignment:ratt:1.0.3:RefSeq:NC_000962.3\n",
      "    Key: locus_tag, Value: ErdFL_0477\n",
      "    Key: note, Value: Simliar to Rv0456B of M. tuberculosis H37Rv;ATT start inherited from H37Rv\n",
      "    Key: product, Value: ['antitoxin MazE1']\n",
      "    Key: transl_table, Value: ['11']\n",
      "    Key: translation, Value: ITTYYYVLLSVTTWVGLRHEAKRELVYRGRRSIGRMPREWACRRSRRFAANGVDAAR*\n",
      "\n",
      "Start;\n",
      "type: CDS\n",
      "location: [837211:837739](+)\n",
      "qualifiers:\n",
      "    Key: codon_start, Value: ['1']\n",
      "    Key: inference, Value: alignment:ratt:1.0.3:RefSeq:NC_000962.3\n",
      "    Key: locus_tag, Value: ErdFL_0778\n",
      "    Key: note, Value: Simliar to Rv0742 of M. tuberculosis H37Rv;ATA start inherited from H37Rv\n",
      "    Key: product, Value: ['hypothetical protein']\n",
      "    Key: transl_table, Value: ['11']\n",
      "    Key: translation, Value: ISFVIAAPEAIAAAATDLASIGSTIGAANAAAAANTTAVLAAGADQVSVAIAAAFGAHGQAYQALSAQAATFHIQFVQALTAGAGSYAAAEAASAASITSPLLDAINAPFLAALGRPLIGNGADGAPGTGAAGGAGGLLFGNGGAGGSGAPGGAGGLLFGNGGAGGPGASGGALG*\n",
      "\n",
      "Start;\n",
      "type: CDS\n",
      "location: [3179110:3179359](+)\n",
      "qualifiers:\n",
      "    Key: codon_start, Value: ['1']\n",
      "    Key: gene, Value: ['vapB23']\n",
      "    Key: inference, Value: alignment:ratt:1.0.3:RefSeq:NC_000962.3\n",
      "    Key: locus_tag, Value: ErdFL_2980\n",
      "    Key: note, Value: Simliar to Rv2862A of M. tuberculosis H37Rv;ATC start inherited from H37Rv\n",
      "    Key: product, Value: ['antitoxin VapB23']\n",
      "    Key: transl_table, Value: ['11']\n",
      "    Key: translation, Value: ILSDEEREAFRQQAAAQQMSLSNWLRQAGLRQLEAQRQRPLRTAQELREFFASRPDETGAEPDWQAHLQVMAESRRRGLPAP*\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for feature in new_record.features:\n",
    "    if feature.type == 'CDS':\n",
    "        if 'pseudo' not in feature.qualifiers:\n",
    "            aa = feature.qualifiers['translation']\n",
    "            problem = ''\n",
    "            \n",
    "            if '*' in aa[:-1]:\n",
    "                problem += 'Pseudo;'\n",
    "            \n",
    "            if aa[0] not in ['M', 'V', 'L']:\n",
    "                problem += 'Start;'\n",
    "            \n",
    "            if aa[-1] != '*':\n",
    "                problem += 'End;'\n",
    "                \n",
    "            if problem != '':\n",
    "                print(problem)\n",
    "                print(feature)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0452741c-f0bc-475e-a2d0-9dbd33ae137f",
   "metadata": {},
   "source": [
    "This are the three cases where the I start codon is inherited from H37Rv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb5cd79e-29d4-403e-9151-0f9abd091795",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ca34880-a6d0-4c91-8acf-e807f5754d8b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "a16be808-e992-4358-884b-1932ec28e8f9",
   "metadata": {},
   "source": [
    "## Peak at the summary TSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "id": "2c9a390b-8fbe-4437-883b-9c9023a9e513",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict"
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(feature.qualifiers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "id": "48b82c34-21ec-4cdc-bd42-073a6648e3be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>gene</th>\n",
       "      <th>type</th>\n",
       "      <th>rv_equivalent</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>source</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>source</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "      <td>4416075</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ErdFL_0001</td>\n",
       "      <td>dnaA</td>\n",
       "      <td>gene</td>\n",
       "      <td>Rv0001</td>\n",
       "      <td>0</td>\n",
       "      <td>1524</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0001 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ErdFL_0001</td>\n",
       "      <td>dnaA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Rv0001</td>\n",
       "      <td>0</td>\n",
       "      <td>1524</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0001 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>rep_origin</td>\n",
       "      <td>None</td>\n",
       "      <td>1524</td>\n",
       "      <td>2051</td>\n",
       "      <td>+</td>\n",
       "      <td>bakta</td>\n",
       "      <td>De novo annotation with bakta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ErdFL_0002</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>gene</td>\n",
       "      <td>Rv0002</td>\n",
       "      <td>2051</td>\n",
       "      <td>3260</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0002 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ErdFL_0002</td>\n",
       "      <td>dnaN</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Rv0002</td>\n",
       "      <td>2051</td>\n",
       "      <td>3260</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0002 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ErdFL_0003</td>\n",
       "      <td>recF</td>\n",
       "      <td>gene</td>\n",
       "      <td>Rv0003</td>\n",
       "      <td>3279</td>\n",
       "      <td>4437</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0003 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ErdFL_0003</td>\n",
       "      <td>recF</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Rv0003</td>\n",
       "      <td>3279</td>\n",
       "      <td>4437</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0003 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ErdFL_0004</td>\n",
       "      <td>ErdFL_0004</td>\n",
       "      <td>gene</td>\n",
       "      <td>Rv0004</td>\n",
       "      <td>4433</td>\n",
       "      <td>4997</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0004 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ErdFL_0004</td>\n",
       "      <td>ErdFL_0004</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Rv0004</td>\n",
       "      <td>4433</td>\n",
       "      <td>4997</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0004 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ErdFL_0005</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>gene</td>\n",
       "      <td>Rv0005</td>\n",
       "      <td>5239</td>\n",
       "      <td>7267</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0005 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ErdFL_0005</td>\n",
       "      <td>gyrB</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Rv0005</td>\n",
       "      <td>5239</td>\n",
       "      <td>7267</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0005 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ErdFL_0006</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>gene</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>7301</td>\n",
       "      <td>9818</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0006 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ErdFL_0006</td>\n",
       "      <td>gyrA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Rv0006</td>\n",
       "      <td>7301</td>\n",
       "      <td>9818</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0006 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ErdFL_0007</td>\n",
       "      <td>ErdFL_0007</td>\n",
       "      <td>gene</td>\n",
       "      <td>Rv0007</td>\n",
       "      <td>9913</td>\n",
       "      <td>10828</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0007 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ErdFL_0007</td>\n",
       "      <td>ErdFL_0007</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Rv0007</td>\n",
       "      <td>9913</td>\n",
       "      <td>10828</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0007 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ErdFL_0008</td>\n",
       "      <td>ileT</td>\n",
       "      <td>gene</td>\n",
       "      <td>Rvnt01</td>\n",
       "      <td>10886</td>\n",
       "      <td>10960</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rvnt01 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ErdFL_0008</td>\n",
       "      <td>ileT</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>Rvnt01</td>\n",
       "      <td>10886</td>\n",
       "      <td>10960</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rvnt01 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ErdFL_0009</td>\n",
       "      <td>alaT</td>\n",
       "      <td>gene</td>\n",
       "      <td>Rvnt02</td>\n",
       "      <td>11111</td>\n",
       "      <td>11184</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rvnt02 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ErdFL_0009</td>\n",
       "      <td>alaT</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>Rvnt02</td>\n",
       "      <td>11111</td>\n",
       "      <td>11184</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rvnt02 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     locus_tag        gene        type rv_equivalent  start      end strand  \\\n",
       "0         None        None      source          None      0  4416075      +   \n",
       "1   ErdFL_0001        dnaA        gene        Rv0001      0     1524      +   \n",
       "2   ErdFL_0001        dnaA         CDS        Rv0001      0     1524      +   \n",
       "3         None        None  rep_origin          None   1524     2051      +   \n",
       "4   ErdFL_0002        dnaN        gene        Rv0002   2051     3260      +   \n",
       "5   ErdFL_0002        dnaN         CDS        Rv0002   2051     3260      +   \n",
       "6   ErdFL_0003        recF        gene        Rv0003   3279     4437      +   \n",
       "7   ErdFL_0003        recF         CDS        Rv0003   3279     4437      +   \n",
       "8   ErdFL_0004  ErdFL_0004        gene        Rv0004   4433     4997      +   \n",
       "9   ErdFL_0004  ErdFL_0004         CDS        Rv0004   4433     4997      +   \n",
       "10  ErdFL_0005        gyrB        gene        Rv0005   5239     7267      +   \n",
       "11  ErdFL_0005        gyrB         CDS        Rv0005   5239     7267      +   \n",
       "12  ErdFL_0006        gyrA        gene        Rv0006   7301     9818      +   \n",
       "13  ErdFL_0006        gyrA         CDS        Rv0006   7301     9818      +   \n",
       "14  ErdFL_0007  ErdFL_0007        gene        Rv0007   9913    10828      +   \n",
       "15  ErdFL_0007  ErdFL_0007         CDS        Rv0007   9913    10828      +   \n",
       "16  ErdFL_0008        ileT        gene        Rvnt01  10886    10960      +   \n",
       "17  ErdFL_0008        ileT        tRNA        Rvnt01  10886    10960      +   \n",
       "18  ErdFL_0009        alaT        gene        Rvnt02  11111    11184      +   \n",
       "19  ErdFL_0009        alaT        tRNA        Rvnt02  11111    11184      +   \n",
       "\n",
       "   source                                        note  \n",
       "0    None                                              \n",
       "1    ratt  Simliar to Rv0001 of M. tuberculosis H37Rv  \n",
       "2    ratt  Simliar to Rv0001 of M. tuberculosis H37Rv  \n",
       "3   bakta               De novo annotation with bakta  \n",
       "4    ratt  Simliar to Rv0002 of M. tuberculosis H37Rv  \n",
       "5    ratt  Simliar to Rv0002 of M. tuberculosis H37Rv  \n",
       "6    ratt  Simliar to Rv0003 of M. tuberculosis H37Rv  \n",
       "7    ratt  Simliar to Rv0003 of M. tuberculosis H37Rv  \n",
       "8    ratt  Simliar to Rv0004 of M. tuberculosis H37Rv  \n",
       "9    ratt  Simliar to Rv0004 of M. tuberculosis H37Rv  \n",
       "10   ratt  Simliar to Rv0005 of M. tuberculosis H37Rv  \n",
       "11   ratt  Simliar to Rv0005 of M. tuberculosis H37Rv  \n",
       "12   ratt  Simliar to Rv0006 of M. tuberculosis H37Rv  \n",
       "13   ratt  Simliar to Rv0006 of M. tuberculosis H37Rv  \n",
       "14   ratt  Simliar to Rv0007 of M. tuberculosis H37Rv  \n",
       "15   ratt  Simliar to Rv0007 of M. tuberculosis H37Rv  \n",
       "16   ratt  Simliar to Rvnt01 of M. tuberculosis H37Rv  \n",
       "17   ratt  Simliar to Rvnt01 of M. tuberculosis H37Rv  \n",
       "18   ratt  Simliar to Rvnt02 of M. tuberculosis H37Rv  \n",
       "19   ratt  Simliar to Rvnt02 of M. tuberculosis H37Rv  "
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_annotations.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0deb823-03a1-4c5a-8329-c89843c514b9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "id": "2d935e7d-7264-4e25-99fc-957e4b225abb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>gene</th>\n",
       "      <th>type</th>\n",
       "      <th>rv_equivalent</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>source</th>\n",
       "      <th>note</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>980</th>\n",
       "      <td>ErdFL_0489</td>\n",
       "      <td>fadB2</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Rv0468</td>\n",
       "      <td>562533</td>\n",
       "      <td>563394</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0468 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>981</th>\n",
       "      <td>ErdFL_0490</td>\n",
       "      <td>umaA</td>\n",
       "      <td>gene</td>\n",
       "      <td>Rv0469</td>\n",
       "      <td>563526</td>\n",
       "      <td>564387</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0469 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982</th>\n",
       "      <td>ErdFL_0490</td>\n",
       "      <td>umaA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Rv0469</td>\n",
       "      <td>563526</td>\n",
       "      <td>564387</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0469 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>983</th>\n",
       "      <td>ErdFL_0491</td>\n",
       "      <td>pcaA</td>\n",
       "      <td>gene</td>\n",
       "      <td>Rv0470c</td>\n",
       "      <td>564486</td>\n",
       "      <td>565350</td>\n",
       "      <td>-</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0470c of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>984</th>\n",
       "      <td>ErdFL_0491</td>\n",
       "      <td>pcaA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Rv0470c</td>\n",
       "      <td>564486</td>\n",
       "      <td>565350</td>\n",
       "      <td>-</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0470c of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>985</th>\n",
       "      <td>ErdFL_0492</td>\n",
       "      <td>ErdFL_0492</td>\n",
       "      <td>gene</td>\n",
       "      <td>Rv0470A</td>\n",
       "      <td>565492</td>\n",
       "      <td>565933</td>\n",
       "      <td>-</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0470A of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>986</th>\n",
       "      <td>ErdFL_0492</td>\n",
       "      <td>ErdFL_0492</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Rv0470A</td>\n",
       "      <td>565492</td>\n",
       "      <td>565933</td>\n",
       "      <td>-</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0470A of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>987</th>\n",
       "      <td>ErdFL_0493</td>\n",
       "      <td>ErdFL_0493</td>\n",
       "      <td>gene</td>\n",
       "      <td>Rv0471c</td>\n",
       "      <td>565863</td>\n",
       "      <td>566352</td>\n",
       "      <td>-</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0471c of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>988</th>\n",
       "      <td>ErdFL_0493</td>\n",
       "      <td>ErdFL_0493</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Rv0471c</td>\n",
       "      <td>565863</td>\n",
       "      <td>566352</td>\n",
       "      <td>-</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0471c of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>989</th>\n",
       "      <td>ErdFL_0494</td>\n",
       "      <td>ErdFL_0494</td>\n",
       "      <td>gene</td>\n",
       "      <td>Rv0472c</td>\n",
       "      <td>566361</td>\n",
       "      <td>567066</td>\n",
       "      <td>-</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0472c of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990</th>\n",
       "      <td>ErdFL_0494</td>\n",
       "      <td>ErdFL_0494</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Rv0472c</td>\n",
       "      <td>566361</td>\n",
       "      <td>567066</td>\n",
       "      <td>-</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0472c of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>991</th>\n",
       "      <td>ErdFL_0495</td>\n",
       "      <td>ErdFL_0495</td>\n",
       "      <td>gene</td>\n",
       "      <td>Rv0473</td>\n",
       "      <td>567202</td>\n",
       "      <td>568573</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0473 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>992</th>\n",
       "      <td>ErdFL_0495</td>\n",
       "      <td>ErdFL_0495</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Rv0473</td>\n",
       "      <td>567202</td>\n",
       "      <td>568573</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0473 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>993</th>\n",
       "      <td>ErdFL_0496</td>\n",
       "      <td>ErdFL_0496</td>\n",
       "      <td>gene</td>\n",
       "      <td>Rv0474</td>\n",
       "      <td>568659</td>\n",
       "      <td>569082</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0474 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>994</th>\n",
       "      <td>ErdFL_0496</td>\n",
       "      <td>ErdFL_0496</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Rv0474</td>\n",
       "      <td>568659</td>\n",
       "      <td>569082</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0474 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>ErdFL_0497</td>\n",
       "      <td>hbhA</td>\n",
       "      <td>gene</td>\n",
       "      <td>Rv0475</td>\n",
       "      <td>569435</td>\n",
       "      <td>570035</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0475 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>ErdFL_0497</td>\n",
       "      <td>hbhA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Rv0475</td>\n",
       "      <td>569435</td>\n",
       "      <td>570035</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0475 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>ErdFL_0498</td>\n",
       "      <td>ErdFL_0498</td>\n",
       "      <td>gene</td>\n",
       "      <td>Rv0476</td>\n",
       "      <td>570146</td>\n",
       "      <td>570410</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0476 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>ErdFL_0498</td>\n",
       "      <td>ErdFL_0498</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Rv0476</td>\n",
       "      <td>570146</td>\n",
       "      <td>570410</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0476 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>ErdFL_0499</td>\n",
       "      <td>ErdFL_0499</td>\n",
       "      <td>gene</td>\n",
       "      <td>Rv0477</td>\n",
       "      <td>570414</td>\n",
       "      <td>570861</td>\n",
       "      <td>+</td>\n",
       "      <td>ratt</td>\n",
       "      <td>Simliar to Rv0477 of M. tuberculosis H37Rv</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      locus_tag        gene  type rv_equivalent   start     end strand source  \\\n",
       "980  ErdFL_0489       fadB2   CDS        Rv0468  562533  563394      +   ratt   \n",
       "981  ErdFL_0490        umaA  gene        Rv0469  563526  564387      +   ratt   \n",
       "982  ErdFL_0490        umaA   CDS        Rv0469  563526  564387      +   ratt   \n",
       "983  ErdFL_0491        pcaA  gene       Rv0470c  564486  565350      -   ratt   \n",
       "984  ErdFL_0491        pcaA   CDS       Rv0470c  564486  565350      -   ratt   \n",
       "985  ErdFL_0492  ErdFL_0492  gene       Rv0470A  565492  565933      -   ratt   \n",
       "986  ErdFL_0492  ErdFL_0492   CDS       Rv0470A  565492  565933      -   ratt   \n",
       "987  ErdFL_0493  ErdFL_0493  gene       Rv0471c  565863  566352      -   ratt   \n",
       "988  ErdFL_0493  ErdFL_0493   CDS       Rv0471c  565863  566352      -   ratt   \n",
       "989  ErdFL_0494  ErdFL_0494  gene       Rv0472c  566361  567066      -   ratt   \n",
       "990  ErdFL_0494  ErdFL_0494   CDS       Rv0472c  566361  567066      -   ratt   \n",
       "991  ErdFL_0495  ErdFL_0495  gene        Rv0473  567202  568573      +   ratt   \n",
       "992  ErdFL_0495  ErdFL_0495   CDS        Rv0473  567202  568573      +   ratt   \n",
       "993  ErdFL_0496  ErdFL_0496  gene        Rv0474  568659  569082      +   ratt   \n",
       "994  ErdFL_0496  ErdFL_0496   CDS        Rv0474  568659  569082      +   ratt   \n",
       "995  ErdFL_0497        hbhA  gene        Rv0475  569435  570035      +   ratt   \n",
       "996  ErdFL_0497        hbhA   CDS        Rv0475  569435  570035      +   ratt   \n",
       "997  ErdFL_0498  ErdFL_0498  gene        Rv0476  570146  570410      +   ratt   \n",
       "998  ErdFL_0498  ErdFL_0498   CDS        Rv0476  570146  570410      +   ratt   \n",
       "999  ErdFL_0499  ErdFL_0499  gene        Rv0477  570414  570861      +   ratt   \n",
       "\n",
       "                                            note  \n",
       "980   Simliar to Rv0468 of M. tuberculosis H37Rv  \n",
       "981   Simliar to Rv0469 of M. tuberculosis H37Rv  \n",
       "982   Simliar to Rv0469 of M. tuberculosis H37Rv  \n",
       "983  Simliar to Rv0470c of M. tuberculosis H37Rv  \n",
       "984  Simliar to Rv0470c of M. tuberculosis H37Rv  \n",
       "985  Simliar to Rv0470A of M. tuberculosis H37Rv  \n",
       "986  Simliar to Rv0470A of M. tuberculosis H37Rv  \n",
       "987  Simliar to Rv0471c of M. tuberculosis H37Rv  \n",
       "988  Simliar to Rv0471c of M. tuberculosis H37Rv  \n",
       "989  Simliar to Rv0472c of M. tuberculosis H37Rv  \n",
       "990  Simliar to Rv0472c of M. tuberculosis H37Rv  \n",
       "991   Simliar to Rv0473 of M. tuberculosis H37Rv  \n",
       "992   Simliar to Rv0473 of M. tuberculosis H37Rv  \n",
       "993   Simliar to Rv0474 of M. tuberculosis H37Rv  \n",
       "994   Simliar to Rv0474 of M. tuberculosis H37Rv  \n",
       "995   Simliar to Rv0475 of M. tuberculosis H37Rv  \n",
       "996   Simliar to Rv0475 of M. tuberculosis H37Rv  \n",
       "997   Simliar to Rv0476 of M. tuberculosis H37Rv  \n",
       "998   Simliar to Rv0476 of M. tuberculosis H37Rv  \n",
       "999   Simliar to Rv0477 of M. tuberculosis H37Rv  "
      ]
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_annotations.head(1000).tail(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc4952ec-4c00-47f7-8dfd-9497acef2c04",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "051d146a-992a-46ab-ac13-f46c1299a857",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c42f4bbc-7029-4366-b7b7-1a233cd3a5b3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00fe335c-18e5-4f3c-945a-c6bcf0139670",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5f2f6c2-ff92-426e-bc83-7b64e428d423",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
